src/terminal/adapter/SixelParser.hpp

```diff
@@ -0,0 +1,125 @@
+
+Module Name:
+- SixelParser.hpp
+
+Abstract:
+- This class handles the parsing of the Sixel image format.
+--*/
+
+#pragma once
+
+#include "til.h"
+#include "DispatchTypes.hpp"
+
+class Cursor;
+class TextBuffer;
+
+namespace Microsoft::Console::VirtualTerminal
+{
+    class AdaptDispatch;
+    class Page;
+    class StateMachine;
+
+    class SixelParser
+    {
+    public:
+        static constexpr VTInt DefaultConformance = 9;
+
+        static til::size CellSizeForLevel(const VTInt conformanceLevel) noexcept;
+        static size_t MaxColorsForLevel(const VTInt conformanceLevel) noexcept;
+
+        SixelParser(AdaptDispatch& dispatcher, const StateMachine& stateMachine, const VTInt conformanceLevel = DefaultConformance) noexcept;
+        void SoftReset();
+        void SetDisplayMode(const bool enabled) noexcept;
+        std::function<bool(wchar_t)> DefineImage(const VTInt macroParameter, const DispatchTypes::SixelBackground backgroundSelect, const VTParameter backgroundColor);
+
+    private:
+        // NB: If we want to support more than 256 colors, we'll also need to
+        // change the IndexType to uint16_t, and use a bit field in IndexPixel
+        // to retain the 16-bit size.
+        static constexpr size_t MAX_COLORS = 256;
+        using IndexType = uint8_t;
+        struct IndexedPixel
+        {
+            uint8_t transparent = false;
+            IndexType colorIndex = 0;
+        };
+
+        AdaptDispatch& _dispatcher;
+        const StateMachine& _stateMachine;
+        const VTInt _conformanceLevel;
+
+        void _parseCommandChar(const wchar_t ch);
+        void _parseParameterChar(const wchar_t ch);
+        int _applyPendingCommand();
+        void _executeCarriageReturn() noexcept;
+        void _executeNextLine();
+        void _executeMoveToHome();
+
+        enum class States : size_t
+        {
+            Normal,
+            Attributes,
+            Color,
+            Repeat
+        };
+        States _state = States::Normal;
+        std::vector<VTParameter> _parameters;
+
+        bool _initTextBufferBoundaries();
+        void _initRasterAttributes(const VTInt macroParameter, const DispatchTypes::SixelBackground backgroundSelect) noexcept;
+        void _updateRasterAttributes(const VTParameters& rasterAttributes);
+        void _scrollTextBuffer(Page& page, const int scrollAmount);
+        void _updateTextCursor(Cursor& cursor) noexcept;
+
+        const til::size _cellSize;
+        bool _displayMode = true;
+        til::rect _textMargins;
+        til::point _textCursor;
+        bool _textCursorWasVisible;
+        til::CoordType _availablePixelWidth;
+        til::CoordType _availablePixelHeight;
+        til::CoordType _maxPixelAspectRatio;
+        til::CoordType _pixelAspectRatio;
+        til::CoordType _sixelHeight;
+        til::CoordType _segmentHeight;
+        til::CoordType _pendingTextScrollCount;
+        til::size _backgroundSize;
+        bool _backgroundFillRequired;

```

Comment:
It seems that these members all have undefined values after construction. Should we initialize them here? (Some more below).

Reply:
Looking through my comments again, I think this is the only one where I'd prefer waiting for your response. All the other ones were just my thoughts while I was reading the code.

Reply:
Yeah, I wasn't really sure what to do here. All of the values that aren't initialized here are ones that are required to be reinitialized on every `DefineImage` call (the initialization is handled in the various `_initXXX` methods). So anything we initialize here would be meaningless - it's never going to be used.

At one point I just had everything set to 0, but that seemed confusing, because you start wondering what it means to have a zero aspect ratio, or a zero sixel height. And the answer is that it doesn't mean anything because those fields aren't supposed to be initialized there!

So in the end I figured it might make more sense to just leave them unset. But if you think it's better to clear all the fields, I'm happy to do that. I don't feel strongly about it either way.

Reply:
I think it makes sense to initialize them. Assuming we do have a bug somewhere, perhaps only in the future if we modify this code, this would allow us to consistently reproduce the bug whereas without the initialization the member values would all be random.